### ISSUE TITLE
Invalid request on any endpoint which requires hawk authentication

### ISSUE THREAD

**Issue by vityafx** [mozilla/fxa-auth-server#1475]
The server throws strange 401 error which I can see in it's log as:

>  fxa-auth-server.INFO: request.summary {"op":"request.summary","code":401,"errno":109,"rid":"1474964165351:msk1wst010:24068:itl7k9v0:10001","path":"/v1/session/status","lang":"en-US,*","agent":"Mozilla/5.0","remoteAddressChain":["127.0.0.1"],"t":14,"uid":"00"}

I make a request from Qt5's network access manager and  the only thing I may get from it is:

```
QDEBUG :  HTTP Code:  401
QDEBUG :  HTTP Message:  "Unauthorized"
QDEBUG :  Qt error message:  "Host requires authentication"
```

Here is an example of `GET` request to `/v1/session/status` with my local server. I can't see any message in the body of answer so I can't know where is the problem. 
```
GET /v1/session/status HTTP/1.1
Content-Type: application/json
Authorization: Hawk id="c834f3ae60e2ffeb02b1c08f0e0e346a316bb99c9fd11797cd989c153ece7418", ts="1474964165", nonce="176b70", mac="qeaVcUnZnGU8Tryrlf7hPDIRUX+DOq2vdUyayXwYfpA="
Connection: Keep-Alive
Accept-Encoding: gzip, deflate
Accept-Language: en-US,*
User-Agent: Mozilla/5.0
Host: localhost:9000

HTTP/1.1 401 Unauthorized
Timestamp: 1474964165
vary: origin,accept-encoding
content-type: application/json; charset=utf-8
strict-transport-security: max-age=15552000; includeSubDomains
x-frame-options: DENY
x-xss-protection: 1; mode=block
x-download-options: noopen
x-content-type-options: nosniff
cache-control: no-cache
content-encoding: gzip
Date: Tue, 27 Sep 2016 08:16:05 GMT
Connection: keep-alive
Transfer-Encoding: chunked
```
I've tested my crypt functions to be correct by `Test Vectors` of both fxa-auth-server and hawk reference, everything seems to be okay. So I just don't know what can be the problem.

I am also unable to set `trace` logging level: `README` says that in order to change logging level you must change the setting in `config.json` file but here is no file like that, here is `config/index.js` file instead, but any change in it has no result. Trying to set `LOG_LEVEL` environment variable also does not help. So I decided to put it into the `packages.json` file in `start` section but even that does not change logging mode. I am stuck.

**Comment by rfk** [comment]
Hi @vityafx!  The "errno: 109" in the log message indicates that this is an error due to invalid request signature: 

  https://github.com/mozilla/fxa-auth-server/blob/master/lib/error.js#L24

So my first guess would be a problem with calculation of the hmac signature.

Would you mind saying a few words about what sort of application you're developing here, and whether you intend to integrate with the mozilla-hosted API?  We typically consider the auth-server API to be an "internal" API which, despite being openly accessible to the public, might change from time to time as the system evolves.  I may be able to advise on a more stable/supported way of doing what you want.

**Comment by vityafx** [comment]
Can you please correct if I am wrong somewhere here?

For making a `GET` request to `/v1/session/status` we must:
1. Obtain `sessionToken` somewhere, for example from `signup`.
2. Create 2 keys (tokenID and reqHMACkey) from `sessionToken` by using `hkdf(sessionToken, Crypt::KW("sessionToken"), 64)`. First 32 bytes is `tokenID`, last is `reqHMACkey`.
3. Generate `nonce` (just a random), unix timestamp (`ts` key), and `normalized string`, which is (for example):

        hawk.1.header\n1474969249\n7a8e0b\nGET\n/v1/session/status\nlocalhost\n9000\n
    
    where first number is a unix timestamp, then `nonce` parameter, then request method, then the api endpoint (url), a host, a port.

4. Calculate `mac` by using this `normalized string` as data and obtained `reqHMACkey` as key: `var mac = hmac_sha256(normalizedString, reqHMACkey)`.
5. Encode calculated `mac` with `Base64`, let's name it `mac_base64`
6. Create a header: `Authorization: Hawk id="tokenID", ts="1474969249", nonce="7a8e0b", mac="mac_base64"`.
7. Use it as request header.

**Comment by rfk** [comment]
>     hawk.1.header\n1474969249\n7a8e0b\nGET\n/v1/session/status\nlocalhost\n9000\n

IIRC this needs two additional `\n` on the end, the fields in this list are supposed to be: hawk header, timestamp, nonce, method, path, host, port, payload hash, extra data.  Any of those fields that's missing or empty should be represented by a blank line.

(Sanity-checked against https://github.com/hueniverse/hawk#protocol-example)

**Comment by vityafx** [comment]
@rfk I tried that yesterday and just now - still have the same error. Should I blame my `hmac_sha256` or `base64` implementation? Is everything else correct in my algorithm?

**Comment by vityafx** [comment]
Probably I know where is the problem.

At first, I've already corrected some mismatches in my implementation and the server:

1. Client's normalized string had `localhost` while server had `127.0.0.1`.
2. Client's normalized string had `\n\n\n\n` in  the end while server had `\n\n\n`.

Now I am able to see that my hmac is different from yours but I don't understand how that is possible because all my crypt functions and hawk-implementation is tested many times against `test vectors` of both `fxa-auth-server` and `hawk-reference`.

**Comment by vityafx** [comment]
Finally found it. Here is my story:

I have debugged the code above and also wrote my own `test.js` with `createHmac`. If I pass a string instead of `credentials.key` here everything works okay but in fxa-auth-server `credentials.key` is of type `Buffer` and half-sized of string (32 vs 64). So here were 2 solutions:

1. Change fxa-auth-server code to use a string instead of `Buffer` in `credentials.key`.
2. Change my code to use `byte`-type array as key in my `hmac_sha256` function. It worked with a string before.

So I did second, changed my code to work with byte array instead of a hex-string. But here is interesting fact - my code was good for hawk-reference tests so I guess there is a problem too.
